### DISCUSSION TITLE
Where are the abbreviations for setExpirationTime's input string documented?

### DISCUSSION THREAD

**Discussion by smohadjer** [panva/jose#618] [Q&A] [answered]
I was wondering if setExpirationTime('2m') means two months or two minutes and had to dig into source code to find out.

**Comment by panva** [comment]
It would appear they are not.

**Reply by panva** [reply]
I have an update ready to go that extends this syntax a little bit, so i'd like to document it properly with the update. However, it's hard to put this into a concise documentation

## Number

If a number is passed as an argument, it's used in the output directly as a NumericDate timestamp.

## Date

If a Date object is passed, the value represented by that Date object is used in the output as a NumericDate timestamp.

## String

When a string is passed, the function recognizes various formats:

### Numeric string formats

Examples: '10s', '-10s', '+10s'  
Accepted units: seconds (sec, secs, second, seconds, s), minutes (minute, minutes, min, mins, m), hours (hour, hours, hr, hrs, h), days (day, days, d), weeks (week, weeks, w).  
The numeric value represents the time duration.

### Time direction string formats

Examples: '10s ago', '10s from now'  
Accepted units: same as above (seconds, minutes, hours, days, weeks).  
The numeric value represents the time duration.  
'ago' or 'from now' specifies the direction of time from the current moment.

**Reply by big-kahuna-burger** [reply]
I'd like to give it a shot of producing a concise .md this weekend based on the comment above + source. Unless you have something already planned @panva ?

**Reply by panva** [reply]
Try it, it needs to go to the `src/jwt/produce.ts` source file for each of the 3 timestamp methods. iat also has an undefined (the default) case btw.

I think it may be worth to try with and without function overload definition.

**Reply by smohadjer** [reply]
For strings having units for month and year will be useful. I thought y is already an accepted unit for year.

**Reply by panva** [reply]
I'm not allowing month/year because they're not exactly convertable to a fixed number of days. A year can be 365/366 depending on the leap year and month can be 28, 29 (leap feb), 30, 31.

**Reply by smohadjer** [reply]
I see. Totally right!

**Reply by big-kahuna-burger** [reply]
But you are allowing for years with a regex and fall-through switch using fractional days. 

You are not allowing months only, which makes sense given the non-fixed nature of months semantically and possible prefix confusions when using 'm' for minutes.

**Reply by panva** [reply]
Ah right. Years are there with a 365.25 days value which is also imprecise even on average since every 100 years there's no leap.

**Reply by big-kahuna-burger** [reply]
> Ah right. Years are there with a 365.25 days value which is also imprecise even on average since every 100 years there's no leap. 

Ikr, it's complete madness when you dive deep into calendar implementations we use and why :)

**Comment by panva** [comment, marked as answer]
Resolved in #619
